rustls-post-quantum/src/hybrid.rs

```diff
@@ -0,0 +1,178 @@
+    post_quantum: Box<dyn ActiveKeyExchange>,
+    name: NamedGroup,
+    layout: Layout,
+    combined_pub_key: Vec<u8>,
+}
+
+impl ActiveKeyExchange for ActiveHybrid {
+    fn complete(self: Box<Self>, peer_pub_key: &[u8]) -> Result<SharedSecret, Error> {
+        let Some((post_quantum_share, classical_share)) = self
+            .layout
+            .split_received_server_share(peer_pub_key)
+        else {
+            return Err(INVALID_KEY_SHARE);
+        };
+
+        let cl = self
+            .classical
+            .complete(classical_share)?;
+        let pq = self
+            .post_quantum
+            .complete(post_quantum_share)?;
+
+        let secret = self
+            .layout
+            .concat(pq.secret_bytes(), cl.secret_bytes());
+        Ok(SharedSecret::from(secret))
+    }
+
+    /// Allow the classical computation to be offered and selected separately.
+    fn hybrid_component(&self) -> Option<(NamedGroup, &[u8])> {
+        Some((self.classical.group(), self.classical.pub_key()))
+    }
+
+    fn complete_hybrid_component(
+        self: Box<Self>,
+        peer_pub_key: &[u8],
+    ) -> Result<SharedSecret, Error> {
+        self.classical.complete(peer_pub_key)
+    }
+
+    fn pub_key(&self) -> &[u8] {
+        &self.combined_pub_key
+    }
+
+    fn ffdhe_group(&self) -> Option<FfdheGroup<'static>> {
+        None
+    }
+
+    fn group(&self) -> NamedGroup {
+        self.name
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct Layout {
+    /// Length of classical key share.
+    pub(crate) classical_share_len: usize,
+
+    /// Length of post-quantum key share sent by client
+    pub(crate) post_quantum_client_share_len: usize,
+
+    /// Length of post-quantum key share sent by server
+    pub(crate) post_quantum_server_share_len: usize,
+
+    /// Whether the post-quantum element comes first in shares and secrets.
+    ///
+    /// For dismal and unprincipled reasons, SECP256R1MLKEM768 has the
+    /// classical element first, while X25519MLKEM768 has it second.
+    pub(crate) post_quantum_first: bool,
+}
+
+impl Layout {
+    fn split_received_client_share<'a>(&self, share: &'a [u8]) -> Option<(&'a [u8], &'a [u8])> {
+        if share.len() != self.classical_share_len + self.post_quantum_client_share_len {
+            return None;
+        }
+
+        Some(match self.post_quantum_first {
+            true => share.split_at(self.post_quantum_client_share_len),
+            false => share.split_at(self.classical_share_len),
+        })
+    }
+
+    fn split_received_server_share<'a>(&self, share: &'a [u8]) -> Option<(&'a [u8], &'a [u8])> {
+        if share.len() != self.classical_share_len + self.post_quantum_server_share_len {
+            return None;
+        }
+
+        Some(match self.post_quantum_first {
+            true => share.split_at(self.post_quantum_server_share_len),
+            false => share.split_at(self.classical_share_len),
+        })
+    }
```

Comment:
Looks like we could deduplicate this logic by extracting `self.post_quantum_{client,server}_share_len`?